### PR TITLE
CDAP-3470 Refactored NamespaceHttpHandler to throw exceptions and als…

### PR DESCRIPTION
**Pull request by sagarkapare**
…o made ServiceUnAvailableException unchecked.

JIRA: https://issues.cask.co/browse/CDAP-3470

1. Updated code in the `NamespaceHttpHandler` to throw exceptions which would be caught by `HttpExceptionHandler` in order to return the appropriate ERROR code.
2. Made the `ServiceUnavailableException` unchecked.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java [hunk 1]
```diff
@@ -67,64 +66,42 @@ public void getAllNamespaces(HttpRequest request, HttpResponder responder) throw
...
-    } catch (NotFoundException e) {
-      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Namespace %s not found", namespaceId));
-    } catch (Exception e) {
-      LOG.error("Failed to read namespace metadata request body.", e);
-      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, e.getMessage());
-    }
+                                        @PathParam("namespace-id") String namespaceId) throws Exception {
+    NamespaceMeta meta = parseBody(request, NamespaceMeta.class);
+    namespaceAdmin.updateProperties(Id.Namespace.from(namespaceId), meta);
+    responder.sendString(HttpResponseStatus.OK, "Properties updated");
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java]: nit: Not your PR, but should we send a more meaningful response here: ``Properties for namespace %s updated``?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java [hunk 2]
```diff
@@ -147,15 +124,13 @@ public void create(HttpRequest request, HttpResponder responder, @PathParam("nam
                            String.format("Namespace '%s' created successfully.", namespaceId));
     } catch (AlreadyExistsException e) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java]: So this is left as is, because we want to keep the API idempotent, right? :+1:

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java [hunk 3]
```diff
@@ -147,15 +124,13 @@ public void create(HttpRequest request, HttpResponder responder, @PathParam("nam
...
     }
   }
 
   @DELETE
   @Path("/unrecoverable/namespaces/{namespace-id}")
-  public void delete(HttpRequest request, HttpResponder responder, @PathParam("namespace-id") String namespace) {
+  public void delete(HttpRequest request, HttpResponder responder, @PathParam("namespace-id") String namespace)
+    throws Exception {
     if (!cConf.getBoolean(Constants.Dangerous.UNRECOVERABLE_RESET, Constants.Dangerous.DEFAULT_UNRECOVERABLE_RESET)) {
       responder.sendString(HttpResponseStatus.FORBIDDEN,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java]: not this PR, but maybe ``HttpExceptionHandler`` should handle ``FORBIDDEN`` as well?

File: cdap-common/src/main/java/co/cask/cdap/common/namespace/AbstractNamespaceClient.java [hunk 4]
```diff
@@ -135,7 +135,9 @@ public void deleteDatasets(Id.Namespace namespaceId) throws Exception {
     if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
       throw new NamespaceNotFoundException(namespaceId);
     } else if (HttpURLConnection.HTTP_FORBIDDEN == response.getResponseCode()) {
-      throw new NamespaceCannotBeDeletedException(namespaceId, response.getResponseBodyAsString());
+      String msg = String.format("Datasets in the namespace '%s' cannot be deleted. ", namespaceId)
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/namespace/AbstractNamespaceClient.java]: Maybe add a ``Reason: `` after ``deleted.``. Also, you can reuse the existing ``String.format`` instead of concatenating the ``response.getResponseBodyAsString()``

**Comment by bdmogal** [comment]
@sagarkapare only minor comments. Rest LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java]: Yes. Will do that as a separate PR with JIRA: https://issues.cask.co/browse/CDAP-3638

File: cdap-common/src/main/java/co/cask/cdap/common/namespace/AbstractNamespaceClient.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-common/src/main/java/co/cask/cdap/common/namespace/AbstractNamespaceClient.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java]: Updated the response.

**Comment by sagarkapare** [comment]
New build is running here - http://builds.cask.co/browse/CDAP-DUT2792-2

**Comment by bdmogal** [comment]
LGTM. Please squash commits before merging.

**Comment by sagarkapare** [comment]
Build passed. Squashed and merging.
